File: pirates-android/app/src/main/java/com/crescentflare/piratesgame/components/basicviews/ButtonView.kt
package com.crescentflare.piratesgame.components.basicviews

import android.annotation.TargetApi
import android.content.Context
import android.content.res.ColorStateList
import android.content.res.Resources
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.graphics.drawable.StateListDrawable
import android.os.Build
import android.support.v4.content.ContextCompat
import android.util.AttributeSet
import android.util.TypedValue
import android.view.Gravity
import com.crescentflare.piratesgame.R
import com.crescentflare.piratesgame.components.utility.CustomThreePatchDrawable
import com.crescentflare.piratesgame.components.utility.ImageSource
import com.crescentflare.piratesgame.components.utility.ViewletUtil
import com.crescentflare.piratesgame.infrastructure.events.AppEvent
import com.crescentflare.piratesgame.infrastructure.events.AppEventLabeledSender
import com.crescentflare.piratesgame.infrastructure.events.AppEventObserver
import com.crescentflare.unilayout.views.UniButtonView
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import java.lang.ref.WeakReference

/**
 * Basic view: a button
 */
class ButtonView : UniButtonView, AppEventLabeledSender {

    // --
    // Statics
    // --

    companion object {

        // --
        // Static: viewlet integration
        // --

        val viewlet: JsonInflatable = object : JsonInflatable {

            override fun create(context: Context): Any {
                return ButtonView(context)
            }

            override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
                if (obj is ButtonView) {
                    // Text
                    obj.text = ViewletUtil.localizedString(
                        obj.context,
                        mapUtil.optionalString(attributes, "localizedText", null),
                        mapUtil.optionalString(attributes, "text", null)
                    )

                    // Font
                    val defaultTextSize = obj.resources.getDimensionPixelSize(R.dimen.buttonText)
                    obj.setTextSize(
                        TypedValue.COMPLEX_UNIT_PX,
                        mapUtil.optionalDimension(attributes, "textSize", defaultTextSize).toFloat()
                    )

                    // Generic view properties
                    ViewletUtil.applyGenericViewAttributes(mapUtil, obj, attributes)

                    // Button background and text color based on color style
                    obj.setColorStyle(ColorStyle.fromString(mapUtil.optionalString(attributes, "colorStyle", "")))

                    // Event handling
                    obj.tapEvent = AppEvent.fromObject(attributes["tapEvent"])

                    // Forward event observer
                    if (parent is AppEventObserver) {
                        obj.eventObserver = parent
                    }
                    return true
                }
                return false
            }

            override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
                return obj is ButtonView
            }

        }


        // --
        // Static: styles
        // --

        fun defaultStyle(): Map<String, Any> {
            return mapOf(
                Pair("padding", listOf(
                    "\$buttonHorizontalPadding",
                    "\$buttonVerticalPadding",
                    "\$buttonHorizontalPadding",
                    "\$buttonVerticalPadding"
                )),
                Pair("minHeight", "\$buttonHeight"),
                Pair("textSize", "\$buttonText")
            )
        }


        // --
        // Static: background generators
        // --

        private fun getPrimaryButtonDrawable(context: Context): Drawable? {
            return generateDrawableForButton(context, "\$secondary", "\$secondaryHighlight")
        }

        private fun getPrimaryHighlightButtonDrawable(context: Context): Drawable? {
            return generateDrawableForButton(context, "\$secondaryHighlight")
        }

        private fun getSecondaryButtonDrawable(context: Context): Drawable? {
            return generateDrawableForButton(context, "\$primary", "\$primaryHighlight")
        }

        private fun getSecondaryHighlightButtonDrawable(context: Context): Drawable? {
            return generateDrawableForButton(context, "\$primaryHighlight")
        }

        private fun getInvertedButtonDrawable(context: Context): Drawable? {
            return generateDrawableForButton(context, "\$inverted", "\$invertedHighlight")
        }

        private fun getInvertedHighlightButtonDrawable(context: Context): Drawable? {
            return generateDrawableForButton(context, "\$invertedHighlight")
        }

        private fun getDisabledButtonDrawable(context: Context): Drawable? {
            return generateDrawableForButton(context, "\$disabled", "\$disabledHighlight")
        }

        private fun getDisabledInvertedButtonDrawable(context: Context): Drawable? {
            return generateDrawableForButton(context, "\$disabledInverted", "\$disabledInvertedHighlight")
        }

        private fun generateDrawableForButton(context: Context, colorDefinition: String, edgeColorDefinition: String? = null): Drawable? {
            val buttonHeight = context.resources.getDimensionPixelSize(R.dimen.buttonHeight).toFloat() / Resources.getSystem().displayMetrics.density
            val drawable: Drawable?
            if (edgeColorDefinition != null) {
                val attributes = mapOf(
                    Pair("type", "generate"),
                    Pair("name", "filledOval"),
                    Pair("width", 64),
                    Pair("height", buttonHeight * 1.625),
                    Pair("imageWidth", 64),
                    Pair("imageHeight", buttonHeight),
                    Pair("color", edgeColorDefinition),
                    Pair("caching", "always"),
                    Pair("otherSources", listOf(mapOf(
                        Pair("type", "generate"),
                        Pair("name", "filledOval"),
                        Pair("width", 52),
                        Pair("height", buttonHeight * 2),
                        Pair("color", colorDefinition)
                    )))
                )
                drawable = ImageSource(attributes).getDrawable(context)
            } else {
                val attributes = mapOf(
                    Pair("type", "generate"),
                    Pair("name", "filledOval"),
                    Pair("width", 64),
                    Pair("height", buttonHeight * 1.625),
                    Pair("imageWidth", 64),
                    Pair("imageHeight", buttonHeight),
                    Pair("color", colorDefinition),
                    Pair("caching", "always")
                )
                drawable = ImageSource(attributes).getDrawable(context)
            }
            if (drawable is BitmapDrawable) {
                val bitmap = drawable.bitmap
                if (bitmap != null) {
                    return CustomThreePatchDrawable(bitmap, (Resources.getSystem().displayMetrics.density * 12).toInt())
                }
            }
            return null
        }

    }


    // --
    // Members
    // --

    private var eventObserverReference: WeakReference<AppEventObserver>? = null


    // --
    // Initialization
    // --

    @JvmOverloads
    constructor(
            context: Context,
            attrs: AttributeSet? = null,
            defStyleAttr: Int = 0)
            : super(context, attrs, defStyleAttr)

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(
            context: Context,
            attrs: AttributeSet?,
            defStyleAttr: Int,
            defStyleRes: Int)
            : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        val buttonHorizontalPadding = resources.getDimensionPixelSize(R.dimen.buttonHorizontalPadding)
        val buttonVerticalPadding = resources.getDimensionPixelSize(R.dimen.buttonVerticalPadding)
        minHeight = 0
        minimumHeight = 0
        minWidth = 0
        minimumWidth = 0
        gravity = Gravity.CENTER
        isAllCaps = false
        setColorStyle(ColorStyle.Primary)
        setPadding(buttonHorizontalPadding, buttonVerticalPadding, buttonHorizontalPadding, buttonVerticalPadding)
        setTextSize(TypedValue.COMPLEX_UNIT_PX, resources.getDimensionPixelSize(R.dimen.buttonText).toFloat())
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            stateListAnimator = null
        }
    }


    // --
    // Configurable values
    // --

    var eventObserver: AppEventObserver?
        get() = eventObserverReference?.get()
        set(newValue) {
            eventObserverReference = if (newValue != null) {
                WeakReference(newValue)
            } else {
                null
            }
        }

    var tapEvent: AppEvent? = null
        set(tapEvent) {
            field = tapEvent
            if (hasOnClickListeners()) {
                setOnClickListener(null)
            }
            if (this.tapEvent != null) {
                setOnClickListener {
                    val currentTapEvent = this@ButtonView.tapEvent
                    if (currentTapEvent != null) {
                        eventObserver?.observedEvent(currentTapEvent, this@ButtonView)
                    }
                }
            }
        }

    fun setColorStyle(colorStyle: ColorStyle) {
        when (colorStyle) {
            ColorStyle.Primary -> {
                background = createStateBackground(
                    getPrimaryButtonDrawable(context),
                    getPrimaryHighlightButtonDrawable(context),
                    getDisabledButtonDrawable(context)
                )
                setTextColor(createStateTitleColor(
                    ContextCompat.getColor(context, R.color.textInverted),
                    ContextCompat.getColor(context, R.color.textInverted),
                    ContextCompat.getColor(context, R.color.textDisabled)
                ))
            }
            ColorStyle.PrimaryInverted -> {
                background = createStateBackground(
                    getInvertedButtonDrawable(context),
                    getInvertedHighlightButtonDrawable(context),
                    getDisabledInvertedButtonDrawable(context)
                )
                setTextColor(createStateTitleColor(
                    ContextCompat.getColor(context, R.color.primary),
                    ContextCompat.getColor(context, R.color.primary),
                    ContextCompat.getColor(context, R.color.primary)
                ))
            }
            ColorStyle.Secondary -> {
                background = createStateBackground(
                    getSecondaryButtonDrawable(context),
                    getSecondaryHighlightButtonDrawable(context),
                    getDisabledButtonDrawable(context)
                )
                setTextColor(createStateTitleColor(
                    ContextCompat.getColor(context, R.color.textInverted),
                    ContextCompat.getColor(context, R.color.textInverted),
                    ContextCompat.getColor(context, R.color.textDisabled)
                ))
            }
        }
    }


    // --
    // Interaction
    // --

    override val senderLabel: String?
        get() = text?.toString()


    // --
    // Helpers
    // --

    private fun createStateBackground(mainDrawable: Drawable?, highlightDrawable: Drawable?, disabledDrawable: Drawable?): Drawable? {
        val drawable = StateListDrawable()
        if (disabledDrawable != null) {
            drawable.addState(intArrayOf(-android.R.attr.state_enabled), disabledDrawable)
        }
        if (highlightDrawable != null) {
            drawable.addState(intArrayOf(android.R.attr.state_selected), highlightDrawable)
            drawable.addState(intArrayOf(android.R.attr.state_focused), highlightDrawable)
            drawable.addState(intArrayOf(android.R.attr.state_pressed), highlightDrawable)
        }
        if (mainDrawable != null) {
            drawable.addState(intArrayOf(), mainDrawable)
        }
        return drawable
    }

    private fun createStateTitleColor(mainColor: Int, highlightColor: Int, disabledColor: Int): ColorStateList {
        return ColorStateList(
            arrayOf(
                intArrayOf(android.R.attr.state_enabled, android.R.attr.state_selected),
                intArrayOf(android.R.attr.state_enabled, android.R.attr.state_focused),
                intArrayOf(android.R.attr.state_enabled, android.R.attr.state_pressed),
                intArrayOf(android.R.attr.state_enabled),
                intArrayOf()
            ),
            intArrayOf(highlightColor, highlightColor, highlightColor, mainColor, disabledColor)
        )
    }


    // --
    // Color style enum
    // --

    enum class ColorStyle(val value: String) {

        Primary("primary"),
        PrimaryInverted("primaryInverted"),
        Secondary("secondary");

        companion object {

            fun fromString(string: String?): ColorStyle {
                for (enum in ColorStyle.values()) {
                    if (enum.value == string) {
                        return enum
                    }
                }
                return Primary
            }

        }

    }

}

File: pirates-ios/PiratesGame/Components/BasicViews/ButtonView.swift
//
//  ButtonView.swift
//  Basic view: a button
//

import UIKit
import UniLayout
import JsonInflator

enum ButtonViewColorStyle: String {
    
    case primary = "primary"
    case primaryInverted = "primaryInverted"
    case secondary = "secondary"
    case navigationBar = "navigationBar"
    case navigationBarInverted = "navigationBarInverted"
    
}

class ButtonView: UniButtonView, AppEventLabeledSender {
    
    // --
    // MARK: Viewlet integration
    // --
    
    class func viewlet() -> JsonInflatable {
        return ViewletClass()
    }
    
    private class ViewletClass: JsonInflatable {
        
        func create() -> Any {
            return ButtonView()
        }
        
        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            if let button = object as? ButtonView {
                // Apply text
                if let localizedText = convUtil.asString(value: attributes["localizedText"]) {
                    button.setTitle(localizedText.localized(), for: .normal)
                } else {
                    button.setTitle(convUtil.asString(value: attributes["text"]), for: .normal)
                }

                // Apply font
                let fontSize = convUtil.asDimension(value: attributes["textSize"]) ?? AppDimensions.buttonText
                button.titleLabel?.font = AppFonts.normal.font(ofSize: fontSize)

                // Generic view properties
                ViewletUtil.applyGenericViewAttributes(convUtil: convUtil, view: button, attributes: attributes)

                // Button background and text color based on color style
                button.setColorStyle(ButtonViewColorStyle(rawValue: convUtil.asString(value: attributes["colorStyle"]) ?? "") ?? .primary)
                
                // Event handling
                button.tapEvent = AppEvent(value: attributes["tapEvent"])

                // Forward event observer
                if let eventObserver = parent as? AppEventObserver {
                    button.eventObserver = eventObserver
                }
                return true
            }
            return false
        }
        
        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return object is ButtonView
        }
        
    }
    
    
    // --
    // MARK: Styles
    // --
    
    class var defaultStyle: [String: Any] {
        get {
            return [
                "padding": [ "$buttonHorizontalPadding", "$buttonVerticalPadding", "$buttonHorizontalPadding", "$buttonVerticalPadding" ],
                "minHeight": "$buttonHeight",
                "textSize": "$buttonText"
            ]
        }
    }
    
    
    // --
    // MARK: Initialization
    // --
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }
    
    fileprivate func setup() {
        setColorStyle(.primary)
        layoutProperties.minHeight = AppDimensions.buttonHeight
        titleLabel?.font = AppFonts.normal.font(ofSize: AppDimensions.buttonText)
        titleLabel?.numberOfLines = 0
    }
    
    
    // --
    // MARK: Configurable values
    // --
    
    weak var eventObserver: AppEventObserver?
    
    var tapEvent: AppEvent? {
        didSet {
            if oldValue != nil {
                removeTarget(self, action: #selector(tappedButton(_:)), for: .touchUpInside)
            }
            if tapEvent != nil {
                addTarget(self, action: #selector(tappedButton(_:)), for: .touchUpInside)
            }
        }
    }
    
    func setColorStyle(_ colorStyle: ButtonViewColorStyle) {
        // Determine colors
        var textColorSet = textColorsForPrimary()
        switch colorStyle {
        case .primary:
            textColorSet = textColorsForPrimary()
            setBackgroundImage(ButtonView.primaryButtonImage, for: .normal)
            setBackgroundImage(ButtonView.primaryHighlightButtonImage, for: .highlighted)
            setBackgroundImage(ButtonView.disabledButtonImage, for: .disabled)
        case .primaryInverted:
            textColorSet = textColorsForPrimaryInverted()
            setBackgroundImage(ButtonView.invertedButtonImage, for: .normal)
            setBackgroundImage(ButtonView.invertedHighlightButtonImage, for: .highlighted)
            setBackgroundImage(ButtonView.disabledInvertedButtonImage, for: .disabled)
        case .secondary:
            textColorSet = textColorsForSecondary()
            setBackgroundImage(ButtonView.secondaryButtonImage, for: .normal)
            setBackgroundImage(ButtonView.secondaryHighlightButtonImage, for: .highlighted)
            setBackgroundImage(ButtonView.disabledButtonImage, for: .disabled)
        case .navigationBar:
            textColorSet = textColorsForNavigationBar()
            setBackgroundImage(nil, for: .normal)
            setBackgroundImage(nil, for: .highlighted)
            setBackgroundImage(nil, for: .disabled)
        case .navigationBarInverted:
            textColorSet = textColorsForNavigationBarInverted()
            setBackgroundImage(nil, for: .normal)
            setBackgroundImage(nil, for: .highlighted)
            setBackgroundImage(nil, for: .disabled)
        }
        
        // Apply title color and padding
        let horizontalPadding = colorStyle == .navigationBar || colorStyle == .navigationBarInverted ? AppDimensions.buttonNavigationHorizontalPadding : AppDimensions.buttonHorizontalPadding
        setTitleColor(textColorSet[0], for: .normal)
        setTitleColor(textColorSet[1], for: .highlighted)
        setTitleColor(textColorSet[2], for: .disabled)
        padding = UIEdgeInsets(top: AppDimensions.buttonVerticalPadding, left: horizontalPadding, bottom: AppDimensions.buttonVerticalPadding, right: horizontalPadding)
    }
    

    // --
    // MARK: Interaction
    // --
    
    var senderLabel: String? {
        get {
            return titleLabel?.text
        }
    }

    @objc func tappedButton(_ sender: Any) {
        if let tapEvent = tapEvent {
            eventObserver?.observedEvent(tapEvent, sender: self)
        }
    }
    

    // --
    // MARK: Helpers
    // --
    
    private func textColorsForPrimary() -> [UIColor] {
        return [ AppColors.textInverted, AppColors.textInverted, AppColors.textDisabled ]
    }
    
    private func textColorsForPrimaryInverted() -> [UIColor] {
        return [ AppColors.primary, AppColors.primary, AppColors.primary ]
    }
    
    private func textColorsForSecondary() -> [UIColor] {
        return [ AppColors.textInverted, AppColors.textInverted, AppColors.textDisabled ]
    }
    
    private func textColorsForNavigationBar() -> [UIColor] {
        return [ AppColors.text, AppColors.text.withAlphaComponent(0.5), AppColors.text.withAlphaComponent(0.5) ]
    }

    private func textColorsForNavigationBarInverted() -> [UIColor] {
        return [ AppColors.textInverted, AppColors.textInverted.withAlphaComponent(0.5), AppColors.textInverted.withAlphaComponent(0.5) ]
    }

    private static var primaryButtonImage: UIImage? {
        get {
            return imageForButton(colorDefinition: "$secondary", edgeColorDefinition: "$secondaryHighlight")
        }
    }
    
    private static var primaryHighlightButtonImage: UIImage? {
        get {
            return imageForButton(colorDefinition: "$secondaryHighlight")
        }
    }
    
    private static var secondaryButtonImage: UIImage? {
        get {
            return imageForButton(colorDefinition: "$primary", edgeColorDefinition: "$primaryHighlight")
        }
    }
    
    private static var secondaryHighlightButtonImage: UIImage? {
        get {
            return imageForButton(colorDefinition: "$primaryHighlight")
        }
    }
    
    private static var invertedButtonImage: UIImage? {
        get {
            return imageForButton(colorDefinition: "$inverted", edgeColorDefinition: "$invertedHighlight")
        }
    }
    
    private static var invertedHighlightButtonImage: UIImage? {
        get {
            return imageForButton(colorDefinition: "$invertedHighlight")
        }
    }
    
    private static var disabledButtonImage: UIImage? {
        get {
            return imageForButton(colorDefinition: "$disabled", edgeColorDefinition: "$disabledHighlight")
        }
    }
    
    private static var disabledInvertedButtonImage: UIImage? {
        get {
            return imageForButton(colorDefinition: "$disabledInverted", edgeColorDefinition: "$disabledInvertedHighlight")
        }
    }

    private class func imageForButton(colorDefinition: String, edgeColorDefinition: String? = nil) -> UIImage? {
        if let edgeColorDefinition = edgeColorDefinition {
            let attributes: [String: Any] = [
                "type": "generate",
                "name": "filledOval",
                "width": 64,
                "height": Float(AppDimensions.buttonHeight) * 1.625,
                "imageWidth": 64,
                "imageHeight": Float(AppDimensions.buttonHeight),
                "color": edgeColorDefinition,
                "caching": "always",
                "otherSources": [
                    [
                        "type": "generate",
                        "name": "filledOval",
                        "width": 52,
                        "height": Float(AppDimensions.buttonHeight) * 2,
                        "color": colorDefinition
                    ]
                ]
            ]
            return ImageSource(dict: attributes).getImage()?.stretchableImage(withLeftCapWidth: 12, topCapHeight: 0)
        }
        let attributes: [String: Any] = [
            "type": "generate",
            "name": "filledOval",
            "width": 64,
            "height": Float(AppDimensions.buttonHeight) * 1.625,
            "imageWidth": 64,
            "imageHeight": Float(AppDimensions.buttonHeight),
            "color": colorDefinition,
            "caching": "always"
        ]
        return ImageSource(dict: attributes).getImage()?.stretchableImage(withLeftCapWidth: 12, topCapHeight: 0)
    }

}
